### PR TITLE
Return error if patch file passed to state file.patch is malformed

### LINKED ISSUE #59805: [BUG] state file.patch pretend to apply malformed patch
**Description**
If patch file provided for `file.patch` state is malformed then state returns `Patch was already applied` but patch is not applied.

```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: True
     Comment: Patch was already applied
     Started: 12:20:50.953163
    Duration: 61.558 ms
     Changes:
```

**Setup**
/srv/salt/example.sls:
```
patch_example:
  file.patch:
    - name: /tmp/example
    - source: salt://example.patch
```

/srv/salt/example.patch:
```
--- /tmp/example        2021-03-15 12:20:20.698463273 +0000
+++ /tmp/example.new    2021-03-15 12:01:26.620433384 +0000
@@ -172,6 +172,7 @@ malformed patch
+EXAMPLE PATCH
+EXAMPLE PATCH
+EXAMPLE PATCH
+EXAMPLE PATCH
```

```
[root@1566f989189e ~]# cat /tmp/example
EXAMPLE FILE TO PATCH
```

/tmp/example
```
EXAMPLE FILE TO PATCH
```

**Steps to Reproduce the behavior**

yum install -y patch
salt-call --local state.sls example -l debug

```
[DEBUG   ] file.managed: {'changes': {'diff': '--- \n+++ \n@@ -0,0 +1,4 @@\n+--- /tmp/example\t2021-03-15 12:20:20.698463273 +0000\n++++ /tmp/example.new\t2021-03-15 12:01:26.620433384 +0000\n+@@ -172,6 +172,19 @@ malformed patch\n++EXAMPLE PATCH\n'}, 'comment': 'File /tmp/__salt.tmp.vkktza9o updated', 'name': '/tmp/__salt.tmp.vkktza9o', 'result': True}
[DEBUG   ] LazyLoaded cmd.run_all
[INFO    ] Executing command /usr/bin/patch in directory '/root'
[ERROR   ] Command '/usr/bin/patch' failed with return code: 2
[ERROR   ] stdout: patching file /tmp/__salt.tmp.pj5ysvms (read from /tmp/example)
[ERROR   ] stderr: /usr/bin/patch: **** unexpected end of file in patch
[ERROR   ] retcode: 2
[INFO    ] Executing command /usr/bin/patch in directory '/root'
[INFO    ] Patch was already applied
```

**Expected behavior**
It would be nice to return error when patch is malformed and can not be applied:
```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: False
     Comment: /usr/bin/patch: **** malformed patch at line 7:
     Started: 12:33:44.915605
    Duration: 59.202 ms
     Changes:
```

**Versions Report**
<details><summary>salt --versions-report</summary>

```
Salt Version:
          Salt: 3002.5

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.6.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: Not Installed
        pygit2: Not Installed
        Python: 3.6.8 (default, Aug 24 2020, 17:57:11)
  python-gnupg: Not Installed
        PyYAML: 3.12
         PyZMQ: 19.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: centos 8
        locale: UTF-8
       machine: x86_64
       release: 3.10.0-1160.15.2.el7.x86_64
        system: Linux
       version: CentOS Linux 8

```
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by msciciel**
### What does this PR do?
If patch file provided for file.patch state is malformed then state
returns `Patch was already applied` but patch is not applied.
It is better to return error in such case.

### What issues does this PR fix or reference?
Fixes: #59805

### Previous Behavior
```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: True
     Comment: Patch was already applied
     Started: 12:20:50.953163
    Duration: 61.558 ms
     Changes:
```

### New Behavior
```
          ID: patch_example
    Function: file.patch
        Name: /tmp/example
      Result: False
     Comment: /usr/bin/patch: **** malformed patch at line 7:
     Started: 12:33:44.915605
    Duration: 59.202 ms
     Changes:
```

Test needs to be fixed also because it return OK(false positive) with original code but log file contains and patch is not applied:
```
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/bar/math.txt b/foo/bar/math.txt
|--- a/foo/bar/math.txt	2018-04-09 18:43:52.883205365 -0500
|+++ b/foo/bar/math.txt	2018-04-09 18:44:58.525061654 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 13
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/numbers.txt b/foo/numbers.txt
|--- a/foo/numbers.txt	2018-04-09 18:43:58.014272504 -0500
|+++ b/foo/numbers.txt	2018-04-09 18:44:46.487905044 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

File: salt/states/file.py [hunk 1]
```diff
@@ -6887,6 +6887,12 @@ def _patch(patch_file, options=None, dry_run=False):
...
 
         pre_check = _patch(patch_file, patch_opts)
+
+        if os.path.getsize(patch_rejects) == 0 and pre_check['retcode'] != 0:
+            ret['comment'] = pre_check['stderr']
+            ret['result'] = False
+            return ret
+
         if pre_check["retcode"] != 0:
             # Try to reverse-apply hunks from rejects file using a dry-run.
```
Comment (danielrobbins) [inline salt/states/file.py:6897]: Just to be super-safe, I would suggest the following change. Move your new code inside the ``if pre_check["retcode"] != 0:`` conditional, immediately after this line. Then change your conditional to be ``if not os.path.exists(patch_rejects) or os.path.getsize(patch_rejects) == 0``. If the user specifies ``reject_file`` then I do not think this file will necessarily exist. The extra ``os.path.exists()`` check will ensure this doesn't result in ``OSError``.

**Review by danielrobbins — CHANGES_REQUESTED** [review]

File: salt/states/file.py (hunk 1 above)
Comment (msciciel) [reply salt/states/file.py:6896]: @danielrobbins fixed. I was trying to provide unit test but I'm lack of knowledge how to mock content of `salt://example.patch` in order to run test for `file.patch`. Please verify if current state is enough or do I need to write a proper unit test.

**Comment by sagetherage** [comment]
@msciciel if you can get this out of a draft status it will run the tests and I can get it reviewed! Thank you!

**Comment by msciciel** [comment]
> @msciciel if you can get this out of a draft status it will run the tests and I can get it reviewed! Thank you!

Done.

**Comment by msciciel** [comment]
Any chance for final review and possible merge?

**Comment by msciciel** [comment]
Test need to be fixed also because it return OK but log file contains:
```
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/bar/math.txt b/foo/bar/math.txt
|--- a/foo/bar/math.txt	2018-04-09 18:43:52.883205365 -0500
|+++ b/foo/bar/math.txt	2018-04-09 18:44:58.525061654 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 13
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -ur a/foo/numbers.txt b/foo/numbers.txt
|--- a/foo/numbers.txt	2018-04-09 18:43:58.014272504 -0500
|+++ b/foo/numbers.txt	2018-04-09 18:44:46.487905044 -0500
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
```

**Comment by msciciel** [comment]
@danielrobbins @sagetherage is there a change for a review of this fix?

**Review by dwoz — APPROVED** [review]

**Pull request merged**